src/main/java/uk/gov/companieshouse/overseasentitiesapi/utils/PublicPrivateDataCombiner.java

```diff
@@ -0,0 +1,142 @@
+package uk.gov.companieshouse.overseasentitiesapi.utils;
+
+import java.security.NoSuchAlgorithmException;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.commons.lang3.tuple.Pair;
+import uk.gov.companieshouse.api.model.beneficialowner.PrivateBoDataApi;
+import uk.gov.companieshouse.api.model.company.CompanyProfileApi;
+import uk.gov.companieshouse.api.model.managingofficerdata.ManagingOfficerDataApi;
+import uk.gov.companieshouse.api.model.officers.CompanyOfficerApi;
+import uk.gov.companieshouse.api.model.psc.PscApi;
+import uk.gov.companieshouse.api.model.update.OverseasEntityDataApi;
+import uk.gov.companieshouse.overseasentitiesapi.service.PrivateDataRetrievalService;
+import uk.gov.companieshouse.overseasentitiesapi.service.PublicDataRetrievalService;
+
+public class PublicPrivateDataCombiner {
+
+  private final PublicDataRetrievalService publicDataRetrievalService;
+  private final PrivateDataRetrievalService privateDataRetrievalService;
+  private final HashHelper hashHelper;
+
+  public PublicPrivateDataCombiner(final PublicDataRetrievalService publicDataRetrievalService,
+      final PrivateDataRetrievalService privateDataRetrievalService, final String salt) {
+    this.privateDataRetrievalService = privateDataRetrievalService;
+    this.publicDataRetrievalService = publicDataRetrievalService;
+
+    this.hashHelper = new HashHelper(salt);
+  }
+
+  /**
+   * Returns an instance of Pair containing the public data on the right and private data on the
+   * left
+   *
+   * @return Pair of the public data on the right and private data on the left
+   */
+  public Pair<CompanyProfileApi, OverseasEntityDataApi> buildMergedOverseasEntityDataPair() {
+    var privateOE = privateDataRetrievalService.getOverseasEntityData();
+    var publicOE = publicDataRetrievalService.getCompanyProfile();
+
+    return Pair.of(publicOE, privateOE);
+  }
+
+  /**
+   * Creates a map of the combined Beneficial Owner data from the public and private APIs <br> Here
+   * the Hashed ID is the key, and the value is a Pair of the public data on the left and the
+   * private data on the right
+   *
+   * @return Map of the combined Beneficial Owner data
+   * @throws NoSuchAlgorithmException
+   */
+  public Map<String, Pair<PscApi, PrivateBoDataApi>> buildMergedBeneficialOwnerDataMap()
+      throws NoSuchAlgorithmException {
```

Comment:
Maybe we should catch NoSuchAlgorithmException in this method and throw ServiceException? May make it easier for callers.